diff --git a/configure.ac b/configure.ac
index 26eef70faa..6e7cb014cb 100644
--- a/configure.ac
+++ b/configure.ac
@@ -89,11 +89,11 @@ if test "x$INTERNAL_OPALDIR" != "x"; then
    datarootdir="\${exec_prefix}"
    makedir="\${datarootdir}/make"
    VERSION_DIR="$INTERNAL_OPALDIR"
-   CPPFLAGS="${CPPFLAGS} -I\${includedir}"
+   CPPFLAGS="${CPPFLAGS} -I\${includedir} -std=c++03"
 else
    makedir="\${datarootdir}/opal/make"
    VERSION_DIR=`pwd`
-   CPPFLAGS="${CPPFLAGS} -I\${includedir}/opal"
+   CPPFLAGS="${CPPFLAGS} -I\${includedir}/opal -std=c++03"
 fi
 
 AC_SUBST(makedir)
diff --git a/make/config.guess b/make/config.guess
index 500ee74b04..1fcd267c37 100644
--- a/make/config.guess
+++ b/make/config.guess
@@ -221,7 +221,14 @@ case "${UNAME_MACHINE}:${UNAME_SYSTEM}:${UNAME_RELEASE}:${UNAME_VERSION}" in
     mvmeppc:OpenBSD:*:*)
 	echo powerpc-unknown-openbsd${UNAME_RELEASE}
 	exit 0 ;;
-    pmax:OpenBSD:*:*)
+    aarch64:Linux:*:*)
+	echo ${UNAME_MACHINE}-unknown-linux-gnu
+	exit ;;
+    aarch64_be:Linux:*:*)
+	UNAME_MACHINE=aarch64_be
+	echo ${UNAME_MACHINE}-unknown-linux-gnu
+	exit ;;
+   pmax:OpenBSD:*:*)
 	echo mipsel-unknown-openbsd${UNAME_RELEASE}
 	exit 0 ;;
     sgi:OpenBSD:*:*)
diff --git a/src/opal/manager.cxx b/src/opal/manager.cxx
index abaced3feb..ac8578d4e3 100644
--- a/src/opal/manager.cxx
+++ b/src/opal/manager.cxx
@@ -158,7 +158,7 @@ PString OpalGetVersion()
 #define BetaCode    "beta"
 #define ReleaseCode "."
 
-  return psprintf("%u.%u%s%u (svn:%u)", MAJOR_VERSION, MINOR_VERSION, BUILD_TYPE, BUILD_NUMBER, SVN_REVISION);
+  return psprintf("%u.%u%s%u (svn:%u)", MAJOR_VERSION, MINOR_VERSION, BUILD_TYPE, BUILD_NUMBER, 1);
 }
 
 
